designer/server/src/main/scala/pl/touk/nussknacker/ui/process/deployment/reconciliation/ScenarioDeploymentReconciler.scala

```diff
@@ -51,4 +87,173 @@ class ScenarioDeploymentReconciler(
     }
   }
 
+  def recoverNotRunningDeploymentsThatShouldBeRunning(shouldRecover: JobsRecoverySettings => Boolean): Future[Unit] = {
+    implicit val user: LoggedUser = NussknackerInternalUser.instance
+    val processingTypeForWhichJobsShouldBeRecovered = processingTypeServicesProvider.all.toList.collect {
+      case (processingType, processingTypeServices) if shouldRecover(processingTypeServices.jobsRecoverySettings) =>
+        processingType
+    }
+    for {
+      notFinishedDeploymentsThatAreNotRunning <- collectNotFinishedDeploymentsThatAreNotRunning(
+        processingTypeForWhichJobsShouldBeRecovered
+      )
+      _ = logRecoveryBegin(notFinishedDeploymentsThatAreNotRunning)
+      runDeploymentCommandsByProcessingType <-
+        Future
+          .sequence(notFinishedDeploymentsThatAreNotRunning.map { case (scenario, deploymentId) =>
+            prepareRunDeploymentCommandForValidScenario(scenario, deploymentId)
+          })
+          .map(_.flatten)
+      recoveryResult <- Future.sequence(runDeploymentCommandsByProcessingType.map {
+        case (processingType, deployCommand) =>
+          recoverScenarioJob(processingType, deployCommand)
+      })
+      _ = logRecoveryEnd(recoveryResult)
+    } yield ()
+  }
+
+  private def logRecoveryBegin(
+      notFinishedDeploymentsThatAreNotRunning: List[(ScenarioWithDetailsEntity[CanonicalProcess], DeploymentId)]
+  ): Unit = {
+    if (notFinishedDeploymentsThatAreNotRunning.isEmpty) {
+      logger.info(
+        s"No jobs to recover."
+      )
+    } else {
+      logger.info(
+        s"Starting jobs recovery process. ${notFinishedDeploymentsThatAreNotRunning.size} jobs to recover."
+      )
+    }
+  }
+
+  private def logRecoveryEnd(
+      recoveryResult: List[Try[Unit]]
+  ): Unit = {
+    if (recoveryResult.nonEmpty) {
+      val (successes, failures) = recoveryResult.partition(_.isSuccess)
+      logger.info(
+        s"Jobs recovery process finished. $successes jobs recovered successfully, $failures jobs recovery failed."
+      )
+    }
+  }
+
+  private def collectNotFinishedDeploymentsThatAreNotRunning(
+      processingTypeForWhichJobsShouldBeRecovered: Iterable[ProcessingType]
+  )(implicit user: LoggedUser): Future[List[(ScenarioWithDetailsEntity[CanonicalProcess], DeploymentId)]] = {
+    implicit val freshnessPolicy: DataFreshnessPolicy = Fresh
+    for {
+      // In the perfect World we would fetch all deployments that are not finished for all scenarios, but with deployments modelled as actions, we have to fetch deployed scenarios
```

Comment:
This logic may looks strange, but to make it easier, it requires to clean-up "actions" which is not so easy. I did some step towards this in previous changes, but it hasn't been cleaned completely so far.